web/srv/api_handlers.go

```diff
@@ -358,3 +360,46 @@ func (h *handler) handleAPICheck(w http.ResponseWriter, req *http.Request, p htt
 		"results": results,
 	})
 }
+
+func (h *handler) handleAPIResourceDefinition(w http.ResponseWriter, req *http.Request, p httprouter.Params) {
+	namespace := p.ByName("namespace")
```

Comment:
We also should add in some validation that captures the case where we might supply empty strings in the params.

Reply:
Hi @dadjeibaah,

Thanks for your feedback :)

There is a reasoning behind it, yes. IMHO the route using the path params does a better job explicitly identifying the concrete resource. Parameters specified in the path are not optional and this route makes that clear to the reader without requiring to dig deeper in the code, making it easier to understand how to use the API. To me, query parameters would be a better fit for filtering, sorting or any other *optional* parameters. In the other API routes, query parameters are used mostly for filtering, not to identify a particular resource, so they are a great match. I didn't see it as a consistency issue as to me they were different use cases. And as a bonus, we avoid the extra validation required for empty parameters as shown in the diff.

There are quite a few routes using path parameters in [server.go](https://github.com/linkerd/linkerd2/blob/65d5778b9393cd41c001b5023cf02a0d7626d03c/web/srv/server.go#L125-L152) so I thought it'd be ok to use it if the use case was appropriate.

Said that, I'm happy to change it if you still think it's more correct the other way.

Thanks!

Reply:
@tegioz, thanks for the explanation, I totally agree that using path parameters makes more sense here and that we can save an extra step in validation by using path parameters instead. I think if we were redesigning the URL scheme for the dashboard, this would be a great opportunity to also change all other `handleAPI...` methods to use path parameters as well. 

However, my biggest concern is that since all `webapp` routes are path parameters and all API related paths use query parameters, we could potentially deviate from this pattern which may lead to more confusion when refactoring/adding to `server.go` down the road. IMO, I think we should keep things consistent and use the query param style as suggested.

Reply:
No problem, done 👍